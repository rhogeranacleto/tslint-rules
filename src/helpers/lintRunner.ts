import { Configuration, Linter, LintResult } from 'tslint';

interface IMaria {
	src: string;
	// tslint:disable-next-line:no-any
	rule: any;
	fileName?: string;
	fix?: boolean;
}

export function lintRunner(data: IMaria) {

	const { src, rule, fileName = `/tmp/test-${Date.now()}.snap`, fix = false } = data;

	const linter = new Linter({ fix });

	linter.lint(
		fileName,
		src,
		Configuration.parseConfigFile({
			rules: {
				[rule.name || rule]: [true]
			},
			rulesDirectory: 'src'
		})
	);

	return linter.getResult();
}

export function getReplacements(result: LintResult) {

	if (result.fixes) {

		return result.fixes.map(fix => {

			const fixJson = fix.toJson().fix;

			if (fixJson) {

				if (Array.isArray(fixJson)) {

					return fixJson.map(replace => replace.innerText);
				}

				return [fixJson.innerText];
			}
		});
	}
}