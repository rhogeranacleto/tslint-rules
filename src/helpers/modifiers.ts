export function removeLastEmptyLineOfTheBlock(block: string) {

	const lines = block.split('\n');
	const bracket = lines.pop();

	for (let i = lines.length - 1; i > 0; i--) {

		if (lines[i].trim()) {

			break;
		}

		lines.pop();
	}

	lines.push(bracket as string);

	return lines.join('\n');
}