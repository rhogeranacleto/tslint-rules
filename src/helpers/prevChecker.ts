// tslint:disable-next-line:no-implicit-dependencies
import { getPreviousStatement, getPreviousToken } from 'tsutils';
import ts from 'typescript';

export function prevStatementChecker(node: ts.Statement, sourceFile: ts.SourceFile) {

	const prev = getPreviousStatement(node);

	if (prev) {

		return checkPrevLine(node, sourceFile, prev);
	}

	return false;
}

export function prevTokenChecker(node: ts.Node, sourceFile: ts.SourceFile) {

	const prev = getPreviousToken(node);

	if (prev) {

		return checkPrevLine(node, sourceFile, prev);
	}

	return false;
}

function checkPrevLine(node: ts.Node, sourceFile: ts.SourceFile, prev: ts.Node) {

	const startLine = ts.getLineAndCharacterOfPosition(sourceFile, node.getStart()).line;
	const prevLine = ts.getLineAndCharacterOfPosition(sourceFile, prev.getEnd()).line;

	return prevLine === (startLine - 1);
}