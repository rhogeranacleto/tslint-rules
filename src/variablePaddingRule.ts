import * as Lint from 'tslint';
// tslint:disable-next-line:no-implicit-dependencies
import { getNextStatement, getPreviousStatement } from 'tsutils';
import ts from 'typescript';

export const NEW_LINE_BEFORE = 'Missing blank line before variable declaration';
export const NEW_LINE_AFTER = 'Missing blank line after variable declaration';

class Walker extends Lint.RuleWalker {

	public visitVariableStatement(variable: ts.VariableStatement) {

		const startLine = ts.getLineAndCharacterOfPosition(this.getSourceFile(), variable.getStart()).line;
		const endLine = ts.getLineAndCharacterOfPosition(this.getSourceFile(), variable.getEnd()).line;

		this.checkNextStatement(variable, endLine);

		this.checkPrevStatement(variable, startLine);

		this.checkParent(variable, startLine);

		super.visitVariableStatement(variable);
	}

	private checkNextStatement(variable: ts.VariableStatement, line: number) {

		const next = getNextStatement(variable);

		if (next && next.kind !== ts.SyntaxKind.VariableStatement) {

			const nextLine = ts.getLineAndCharacterOfPosition(this.getSourceFile(), next.getStart()).line;

			if (nextLine === line + 1) {

				const fix = new Lint.Replacement(
					next.getFullStart(),
					next.getFullWidth(),
					`\n${next.getFullText()}`
				);

				this.addFailureAtNode(variable, NEW_LINE_AFTER, fix);
			}
		}
	}

	private checkPrevStatement(variable: ts.VariableStatement, line: number) {

		const prev = getPreviousStatement(variable);

		if (prev && prev.kind !== ts.SyntaxKind.VariableStatement) {

			const prevLine = ts.getLineAndCharacterOfPosition(this.getSourceFile(), prev.getStart()).line;

			if (prevLine === line - 1) {

				const fix = new Lint.Replacement(
					variable.getFullStart(),
					variable.getFullWidth(),
					`\n${variable.getFullText()}`
				);

				this.addFailureAtNode(variable, NEW_LINE_BEFORE, fix);
			}
		}
	}

	private checkParent(variable: ts.VariableStatement, line: number) {

		if (variable.parent.kind !== ts.SyntaxKind.SourceFile) {

			const parentBrackets = variable.parent.getChildAt(0);
			const parentLine = ts.getLineAndCharacterOfPosition(this.getSourceFile(), parentBrackets.getStart()).line;

			if (parentLine === line - 1) {

				const fix = new Lint.Replacement(
					variable.getFullStart(),
					variable.getFullWidth(),
					`\n${variable.getFullText()}`
				);

				this.addFailureAtNode(variable, NEW_LINE_BEFORE, fix);
			}
		}
	}
}

export class Rule extends Lint.Rules.AbstractRule {

	public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {

		return this.applyWithWalker(new Walker(sourceFile, this.getOptions()));
	}
}