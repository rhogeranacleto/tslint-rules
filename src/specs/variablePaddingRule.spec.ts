import { getFileText, lintRunner } from '../helpers/lintRunner';
import { NEW_LINE_AFTER, NEW_LINE_BEFORE } from '../variablePaddingRule';

const RULE = 'variable-padding';

describe('maria', () => {

	it.each([
		[`function() {
			const maria = 'maria';
			console.log(maria);
		}`, 2, [NEW_LINE_AFTER, NEW_LINE_BEFORE]],
		[`function() {
			console.log('oi');
			const maria = 'maria';
			console.log(maria);
		}`, 2, [NEW_LINE_AFTER, NEW_LINE_BEFORE]],
		[`import moment from 'moment';
		const maria='joaoa';`, 1, [NEW_LINE_BEFORE]],
		[`export const MARIA = true;`, 0, []],
		[`import moment from 'moment';

		const maria='joaoa';`, 0, []],
		[`const maria = true;
		const pedro = false;

		console.log(maria, pedro);`, 0, []],
		[`function maria() {

			const objeto = {
				um_objeto: 3,
				outra_key: 2
			};
			chamaAFuncao();
		}`, 1, [NEW_LINE_AFTER]]
	])('%s', (src: string, count: number, errors: string[]) => {

		const result = lintRunner({
			src,
			rule: RULE
		});

		expect(result.errorCount).toBe(count);

		expect(result.failures.map(f => f.getFailure())).toEqual(errors);
	});

	it.each([
		[`function() {
			const maria = 'maria';
			console.log(maria);
		}`],
		[`function() {
			console.log('oi');
			const maria = 'maria';
			console.log(maria);
		}`],
		[`import moment from 'moment';
		const maria='joaoa';`],
		[`function maria() {

			const objeto = {
				um_objeto: 3,
				outra_key: 2
			};
			chamaAFuncao();
		}`],
		[`import moment from 'moment';
		const maria='joaoa';
		const maria='joaoa';
		console.log('maria')`]
	])('%s', (src: string) => {

		const result = lintRunner({
			src,
			rule: RULE,
			fix: true
		});

		expect(result.errorCount).toBe(0);

		expect(getFileText(result)).toMatchSnapshot();
	});
});