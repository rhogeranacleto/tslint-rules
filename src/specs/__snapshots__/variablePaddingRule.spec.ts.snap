// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`maria function maria() {

			const objeto = {
				um_objeto: 3,
				outra_key: 2
			};
			chamaAFuncao();
		} 1`] = `
"function maria() {

			const objeto = {
				um_objeto: 3,
				outra_key: 2
			};

			chamaAFuncao();
		}"
`;

exports[`maria function() {
			console.log('oi');
			const maria = 'maria';
			console.log(maria);
		} 1`] = `
"function() {
			console.log('oi');

			const maria = 'maria';

			console.log(maria);
		}"
`;

exports[`maria function() {
			const maria = 'maria';
			console.log(maria);
		} 1`] = `
"function() {

			const maria = 'maria';

			console.log(maria);
		}"
`;

exports[`maria import moment from 'moment';
		const maria='joaoa'; 1`] = `
"import moment from 'moment';

		const maria='joaoa';"
`;
